cidr.1
.TH CIDR 1
.SH NAME
cidr \- RFC 1878 subnet helper
.SH SYNOPSIS
.TP
.B cidr [\-v]
.TP
.B cidr <ipaddress> <subnetmask>
.TP
.B cidr <ipaddress/prefix>
.SH DESCRIPTION
.B cidr
is a utility for determining the network address and broadcast address
for a CIDR subnet.  It takes as input an ip address and netmask or 
and ip/prefix combination.  It outputs the network address, broadcast
address, wildcard address and a network size count.
.SH OPTIONS
.TP
.B \-v
print version information
.TP
.B "<ipaddress> <subnetmask>"
IP address will full subnet mask expansion ie. 203.0.113.0 255.255.255.0
.TP
.B "<ipaddress/prefix>"
IP address as an ip/cidr combination ie. 203.0.113.0/24
.SH EXAMPLES

cidr 198.51.100.0/28

cidr 198.51.100.0 255.255.255.240

.SH COPYRIGHT
cidr.c -- determine CIDR network and broadcast address from
IPv4 ip address and netmask

Heavily Modified October 2015 Stephen Rozanc
Copyright Original (C) 2000 Robert L. Lineberger 

This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

.SH AUTHOR
Stephen Rozanc <github.com/krallice>
.SH CONTRIBUTORS
Robert L. Lineberger <robert@geeksoul.com>

David A. Bandel

Iain Lea

Herman Robers
.SH SEE ALSO
RFC 1878


ChangeLog
*******************  VERSION 2.3.2 *********************
More free() troubles.  Thanks to ken_yap atatatat E M A I L dot C O M
for pointing out a segfault.  Took out malloc()/free() calls for both
holdaddress and holdprefix.  strtok() calls assign these pointers right
off anyway.

Thanks to Herman Robers <herman.robers@pinewood.nl>
for the Solaris and RedHat stuff, and the nice Makefile.
 
*******************  VERSION 2.3.1 *********************
doh!
Bugfix in cleanup.  Pointers had been set to NULL before free()'ing.
 
*******************  VERSION 2.3 ***********************

Improvements:

1.  Manual page.

2.  "-v" option to print version information.

3.  modified usage().

Minor bugfixes:

1.  Fixed problem with segfault when only one argument is given and
    that argument is not valid.

2.  Fixed stupid malloc().

*******************  VERSION 2.2 *********************

This version incorporates ideas suggested by Iain Lea, namely:

1. Short form usage, such as `cidr 192.168.1.1/27`
2. Better output format.  IP addresses are aligned.
3. Example uses in usage() output.

I think these are nice improvements and justify this somewhat rapid
update.

Bugs:

1. Doug Denny reports that inet_aton() and inet_ntoa() cause linker
   errors when building cidr 2.0 under Solaris 2.7.

2. ip address 255.255.255.255 / mask 255.255.255.255 seems to hang on
   the host count.  I say "seems" because for some masks there is a long
   wait while it calculates the host count, but those masks are usually
   the ones resulting in a large number of hosts, such as a /8 and above.
   A /32 count is usually instantaneous, but it seems to get stuck when
   used with 255.255.255.255 ip address.

All previous functionality remains.  Long form is still there:
i.e. `cidr -b 11111111000000001111111100000000 -q 255.255.255.255 -H`

*******************  VERSION 2 *****************************

This version incorporates ideas suggested by David A. Bandel:

1. There is now a "-H" option to print host addresses.

2. The host addresses will correctly represent valid addresses for
   unusual netmasks, i.e., masks with non-contiguous 1-bits, such as
   11111111111111111111111100001000

3. The total addresses count now reflects only a count of valid host
   addresses within the given subnet.  In version 1, the network address
   and broadcast address were included in the count.



cidr.c
/*************************************************************************

    cidr.c -- determine CIDR network and broadcast address from
    IPV4 ip address and netmask
    Copyright (C) 2000 Robert L. Lineberger 
    Heavily Modified 2015 Stephen Rozanc.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

***************************************************************************/

#include "cidr.h"
#include "network.h"

const char *version_val="2.4.0";

int main( int argc, char *argv[] ) {

	// Define our network structure:
	network_t network;

	// If not enough (or too many) arguments are supplied, bomb out:
	if ( argc < 2 ) {
		usage(argv[0]);
	} else if ( argc > 3 ) {
		invalid(0, argv[0]);
	// Else if we have exactly 2 arguments (filename + switch):
	} else if ( argc == 2 ) {
		// Check to see if user is requesting version or help information:
		if (strncmp((const char *) argv[1],"-v",2) == 0) {
			print_version();
		} else if (strncmp((const char *) argv[1],"-h",2) == 0) {
			usage(argv[0]);
		// Else end user is entering CIDR notation, let's check:
		} else {
			char *ip;
			char *cidr;
			if ( splitCIDR(&network, argv[1], ip, cidr) != 1 ) {
				invalid(1, argv[1]);
			}
		}
	// End user has entered 2 arguments, hopefully meaning ip + subnet mask:
	} else if ( argc == 3 ) {
		// Check to see if our IP/Masks are valid:
		if ( setIPAddress(&network, argv[1]) != 1 ) {
			invalid(1, argv[1]);
		}
		if ( setSubnetMask(&network, argv[2]) != 1 ) {
			invalid(2, argv[2]);
		}
	}
	printNetworkDetails(&network);
	return 0;
}

// Print Usage Patterns to end user:
void usage(char *arg) {

	printf("\ncidr version %s October, 2015\n", version_val);
	printf("Version 2.4.X Heavily Modified 2015 Stephen Rozanc\n");
	printf("Original - Copyright 2000 Robert L. Lineberger\n\n");
	printf("https://github.com/krallice\n");
	printf("robert@geeksoul.com\n");
	printf("http://geeksoul.com/robert/cidr.html\n");
	printf("License: GPL\n\n");

	printf("Contributors:\n");
	printf("Stephen Rozanc\n");
	printf("David A. Bandel\n");
	printf("Iain Lea\n");
	printf("Herman Robers\n\n");

	printf(
	"Usage:\n\n"
	"%s\t\t\t\tprint usage information (this message)\n"
	"%s [-v]\t\t\tprint version information\n"
	"%s [ipaddress/prefix]\tuse short form CIDR notation ie. 198.51.100.0/24 \n"
	"%s [ipaddress] [subnetmask]\tuse full subnet mask ie. 198.51.100.0 255.255.255.0"
	"     \n\n",arg,arg,arg,arg);

	exit(EXIT_FAILURE);
}

// Print Version information to end user:
void print_version(void) {

	fprintf(stderr,"cidr version %s\n",version_val);
	exit(EXIT_SUCCESS);
}

void invalid(int argnum, char* procpath)
{
 printf("invalid argument: argv[%d]\n",argnum);
 usage(procpath);
 exit(EXIT_FAILURE);
}
